src/storage_stage.rs

```diff
@@ -0,0 +1,450 @@
+        keypair: &Arc<Keypair>,
+        _ledger_path: &str,
+        entry_id: Hash,
+        entry_height: u64,
+    ) -> Result<()> {
+        let mut seed = [0u8; 32];
+        let signature = keypair.sign(&entry_id.as_ref());
+
+        seed.copy_from_slice(&signature.as_ref()[..32]);
+
+        let mut rng = ChaChaRng::from_seed(seed);
+
+        // Regenerate the answers
+        let num_slices = (entry_height / ENTRIES_PER_SLICE) as usize;
+        if num_slices == 0 {
+            info!("Ledger has 0 slices!");
+            return Ok(());
+        }
+        // TODO: what if the validator does not have this slice
+        let slice = signature.as_ref()[0] as usize % num_slices;
+
+        debug!(
+            "storage verifying: slice: {} identities: {}",
+            slice, NUM_IDENTITIES,
+        );
+
+        let mut samples = vec![];
+        for _ in 0..NUM_SAMPLES {
+            samples.push(rng.gen_range(0, 10));
+        }
+        debug!("generated samples: {:?}", samples);
+        // TODO: cuda required to generate the reference values
+        // but if it is missing, then we need to take care not to
+        // process storage mining results.
+        #[cfg(feature = "cuda")]
+        {
+            let mut storage_results = _storage_results.write().unwrap();
+
+            // Lock the keys, since this is the IV memory,
+            // it will be updated in-place by the encryption.
+            // Should be overwritten by the vote signatures which replace the
+            // key values by the time it runs again.
+            let mut storage_keys = _storage_keys.write().unwrap();
+
+            match chacha_cbc_encrypt_file_many_keys(
+                _ledger_path,
+                slice as u64,
+                &mut storage_keys,
+                &samples,
+            ) {
+                Ok(hashes) => {
+                    debug!("Success! encrypted ledger slice: {}", slice);
+                    storage_results.copy_from_slice(&hashes);
+                }
+                Err(e) => info!("error encrypting file: {:?}", e),
+            }
+        }
+        // TODO: bundle up mining submissions from replicators
+        // and submit them in a tx to the leader to get reward.
+        Ok(())
+    }
+
+    pub fn process_entries(
+        keypair: &Arc<Keypair>,
+        storage_keys: &Arc<RwLock<StorageKeys>>,
+        storage_results: &Arc<RwLock<StorageResults>>,
+        entry_receiver: &EntryReceiver,
+        ledger_path: &str,
+        poh_height: &mut u64,
+        entry_height: &mut u64,
+        current_key_idx: &mut usize,
+    ) -> Result<()> {
+        let timeout = Duration::new(1, 0);
+        let entries = entry_receiver.recv_timeout(timeout)?;
+
+        for entry in entries {
+            {
+                // Go through the transactions, find votes, and use them to update
+                // the storage_keys with their signatures.
+                for tx in entry.transactions {
```

Comment:
Are you able to leverage `votes()` for this? https://github.com/solana-labs/solana/blob/master/src/ledger.rs#L498

Reply:
It doesn't give the signature, just (pubkey, vote, hash)